src/app.js

```diff
@@ -31,7 +33,23 @@ io.of('/game').on('connection', (socket) => {
   socket.on(MOVE_PLAYED, async ({ gameId, playerId, move }) => {
     const game = await dynamo.getGameById(gameId)
     const { piece, position } = move
+
+    console.log(isInvalidMove(game, playerId, position))
+
+    if (isInvalidMove(game, playerId, position)) {
+      return
+    }
+
     game.board[position] = piece
+
+    if (isGameFinished(game)) {
```

Comment:
As we discussed in the call, we can do something like `const [isFinished, winner] = isGameFinished(game)`
And then in the `isGameFinished()` do something like:
```js
export const isGameFinished = ({ players, board }) => {
  let isFinished = false
  let winner = null

  for (const line of lines) {
    const [firstPlayer, secondPlayer] = players
    const [a, b, c] = lines

    if (board[a] !== EMPTY && board[a] === board[b] && board[b] === board[c]) {
      isFinished = true
      winner = firstPlayer.piece === board[a] ? firstPlayer.id : secondPlayer.id
      break
    }

    if (stalemate(board)) {
      isFinished = true
      break
    }
  }

  return [isFinished, winner]
}

const stalemate = (board) => {
  for (let i = 0; i < board.length; i++) {
    if (board[i] === EMPTY) return false
  }
  return true
}
```

Reply:
Also  https://reactjs.org/tutorial/tutorial.html 👀